### ISSUE TITLE
BaseActiveRecord::afterConstruct() is missing

### ISSUE THREAD

**Issue by DJLebedev** [yiisoft/yii2#7373]
Why method "afterConstruct()" was deleted from Yii2 and how can I replace it? I need to call "ActiveRecord::loadDefaultValues()" (or some other logic) inside model class (but only for new records, not for existent records).

**Comment by lynicidn** [comment]
use init
but loadDefaultValues not needed for finded record

**Comment by kiaplayer** [comment]
But how can we differ new record from existent record inside "init()"?

**Comment by lynicidn** [comment]
@kiaplayer am it mean too - afterConsrtuct not solve it

**Comment by dapatrese** [comment]
afaik `construct` and `init` are both called too early to detemine if the record is new.

Two possible solutions:

1. You could write your own constructor method and use it instead of `new` when you create new model objects, e.g. `MyModel::create()`.

2. You could override `BaseActiveRecord::instantiate` to prevent constructor from being called when instantiating existing records. @ragazzo suggested two ways to achieve this (see #5786 and #5771). In that case you could simply use the constructor for your needs.

probably depends on your existing code and/or personal taste

**Comment by creocoder** [comment]
Use `init()` for new records and `afterFind()` for existent record. Its **absoluttely** enough for 99,9% cases.

**Comment by lynicidn** [comment]
@creocoder  init for all

**Comment by creocoder** [comment]
@lynicidn 

> upd: how detect in init isNewRecord?

Read upper comment.

**Comment by qiangxue** [comment]
Use `init`. It is like `afterConstruct`.

**Comment by kiaplayer** [comment]
@qiangxue 
But `afterConstruct` in Yii1 is called only after creating new models (not after getting models from DB).
`init` in Yii2 is called in both cases.

**Comment by qiangxue** [comment]
The Yii1 AR construct is a bad implementation. We do not want to take that route again.
If general, if you want to do something specific for new models, do it in `beforeSave()`. If you want to do it earlier (e.g. calling `loadDefaultValues()`), it's recommended that you do it outside of AR.

**Comment by DJLebedev** [comment]
Thank you @kiaplayer for expressing my thoughts.
@qiangxue 
In this case, I consider it inappropriate advice:  https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveRecord.php#L103
If you find 1000 records with 10 relations (render grid for exapmple), method loadDefaultValues() will be called 11000 times, absolutely useless.

**Comment by lynicidn** [comment]
`loadDefaultValues` used before **create**

**Comment by DJLebedev** [comment]
@lynicidn, I don`t understand you...

**Comment by lynicidn** [comment]
@DJLebedev loadDefaultValues should be applied **only actionCreate**. In your sample you want use it in **find action**

**Comment by DJLebedev** [comment]
@lynicidn That's it. I don`t want! This is idea of @cebe, judging by PHPDOC.

**Comment by dapatrese** [comment]
What about this workaround?

We introduce a new property `$isExistingRecord` and override `instantiate` to set it. In `init` it can be used to determine if record is new or existing. This could be done in the model itself or in a baseclass which our models extend. Should work well without touching framework code.

```php
class MyModel extends ActiveRecord
{
	public $isExistingRecord = false;

	public static function instantiate($row)
	{
		return new static ([ 'isExistingRecord' => true ]);
	}

	public function init()
	{
		parent::init();
		if ( $this->isExistingRecord )
		{
			// existing record
		}
		else
		{
			// new record
		}
	}
}
```

**Comment by cebe** [comment]
I have updated the documentation about init() method, it is not correct.

> If you find 1000 records with 10 relations (render grid for exapmple), method loadDefaultValues() will be called 11000 times, absolutely useless.

Note that this will not result in 11000 db calls as these are fetched from table schema class, so it is not that bad.

**Comment by DJLebedev** [comment]
@cebe I understand, but I still think it's extra work. Thank you for updating docs.
